### DISCUSSION TITLE
Better permissions generation with trpc-shield

### DISCUSSION THREAD

**Discussion by lewebsimple** [omar-dulaimi/prisma-trpc-generator#66] [Ideas]
First off, thanks for such an amazing project !

Quick question: is there a difference between using `prisma-trpc-shield-generator` vs `prisma-trpc-generator` with `withShield = true` ?

Also, I know this has been discussed elsewhere but I wanted to chime in on possible enhancements to the permissions generation.

## Separating the generated permissions per-model

This would provide better maintainability for large-scale projects, i.e. individual files like `generated/shield/User.shield.ts` could be merged (using something like [unjs/defu](https://github.com/unjs/defu)) and exported from `generated/shield/index.ts`, plus the actual `shield<Context>({...})` function could be called only once from there.

## Preventing the permissions from being overwritten by each generation

Building upon the modular shield files described above, it becomes easier to detect if the corresponding shield file already exists and skip the generation if it's already there.

I don't know if this approach could break or complicate the typing, but it would certainly be an elegant and maintainable solution to permissions generation.

**Comment by omar-dulaimi** [comment]
Hey @lewebsimple 
Glad to hear that you like this project.

To answer your question, this generator basically uses multiple other ones and provides a basic API to interact with them. This includes generators for zod and trpc shield. So instead of installing them separately, using this generator you can use them immediately.

Regarding the enhancements, while they seem quite useful, they come with their own set of challenges. Currently the shield can handle namespaced routers https://github.com/omar-dulaimi/trpc-shield/issues/8 
And even more nesting is being requested as you can see in the issue mentioned. So all this nesting need to be supported by the suggested separation.

As for the second enhancement, I think it would be if it reads then build on top to avoid overwrites. Simple shield generation skipping will be available soon in a currently open PR.

What do you think?
We may have to decide the depth and kinds of nesting to support, and also maybe prepare a POC to make sure all works as expected.
